### DISCUSSION TITLE
Detect different references of same entities as one

### DISCUSSION THREAD

**Discussion by shrinidhin** [explosion/spaCy#10585] [Help: Coding & Implementations] [answered]
Hello!I have built a custom NER using spacy to detect custom entities. It detects all the entities as required. There are certain scenarios of different ways the same entity is referenced.

Scenario 1: Company name and its abbreviations Eg: `Mahindra and Mahindra` also referenced as `M&M`
Scenario 2: For multi word entities, Complete name of company is referenced once in the doc and thereafter it is referenced using the first word. 
`Eg: Tata Consultancy Services reported losses this quarter. Tata released a statement detailing its performance over the last quarter.`
Here `Tata Consultancy Services` and `Tata` are correctly picked out as entities but these are the same entity i.e. both refer to `Tata Consultancy Services`.
Scenario 3: Company having two different names i.e. `Company ABC` and `XYZ` are talking about the same company or entity.

While I understand that the NER component as a whole only recognizes the entities, whether they are same or not is something that has to be handled post processing, I was thinking whether using an Entity ruler will be helpful here?What other approach can I use apart from maintaining a local dictionary with all possible references and checking via code?While the latter seems easier to implement, it seems time consuming to run everytime you pass a doc to detect entities. 
Thank you!

**Comment by thomashacker** [comment, marked as answer]
Hello,
I think `EntityLinking` would be helpful for your usecase (https://spacy.io/api/entitylinker) it connects found entities to unique identifiers.

**Reply by shrinidhin** [reply]
Thank you!

**Reply by shrinidhin** [reply]
@thomashacker Hi!I have been working on implementing Entity Linker for the said use case. I would like to know how to tag the training data in case of sentences with multiple entity labels?Should they all be listed separated by a comma or should there be a single sentence with each entity, similar to the blinding logic used in [Healthsea](https://github.com/explosion/healthsea) ?

**Reply by polm** [reply]
The way training data for the EntityLinker is prepared is to just have entities with the right KB ID set on each of them (and the right KB config). So you should be able to annotate entities as usual and just add KB info. 

I am not sure what you mean about being "separated by a comma".

**Reply by shrinidhin** [reply]
Hi!My question is how do I add KB IDs for each entity in sentences with multiple entities?Consider the following eg:

`Axis Bank, Maruti Suzuki, Biocon can give double-digit returns in short term.`

Here the entities are `Axis Bank,Maruti Suzuki, Biocon`. How will I assign 3 different IDs for each of them?

**Reply by polm** [reply]
KB ID is just a property you can set on each entity span, like this:

```
for ent in doc.ents:
    ent.kb_id_ = ... whatever ...
```

You can't use the script in the demo project as-is because it only has one entity per sentence, but it should be simple to write something that does this if you have an ID associated with each entity somehow.

**Reply by shrinidhin** [reply]
Thank you!Can you help me understand Why are we storing an nlp object to the disk?In my case can I use my trained NER model object as input for the `nlp=spacy.load()` part?

**Reply by polm** [reply]
I'm not really sure what you're referring to, could you link to the code? You should be able to use your own NER model just fine.

**Reply by shrinidhin** [reply]
In the [create_kb.py](https://github.com/explosion/projects/blob/v3/tutorials/nel_emerson/scripts/create_kb.py) file, We are storing the nlp object along with our kb to the disk. Since I have an already trained NER model, already stored locally, i wanted to clarify I can use the same. 
Thank you very much for your help!

**Reply by polm** [reply]
Ah, yes, you can ignore the steps related to saving the nlp object there if you already have a pipeline, since that's just for adding the entity ruler and sentencizer.

**Reply by shrinidhin** [reply]
Great!Thank you!
